lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -3334,3 +3380,114 @@ instance FromJSON (ApiT SmashServer) where
     parseJSON = fromTextJSON "SmashServer"
 instance ToJSON (ApiT SmashServer) where
     toJSON = toTextJSON
+
+{-------------------------------------------------------------------------------
+                         Token minting types
+-------------------------------------------------------------------------------}
+
+-- | Data required when submitting a mint/burn transaction. Cardano implements
+-- minting and burning using transactions, so some of these fields are shared
+-- with @PostTransactionData@.
+data PostMintBurnAssetData (n :: NetworkDiscriminant) = PostMintBurnAssetData
+    { mintBurn   :: !(ApiMintBurnData n)
+    -- ^ Minting and burning request.
+    , passphrase :: !(ApiT (Passphrase "lenient"))
+    -- ^ Passphrase of the wallet.
+    , metadata   :: !(Maybe (ApiT TxMetadata))
+    -- ^ Metadata to attach to the transaction that mints/burns.
+    , timeToLive :: !(Maybe (Quantity "second" NominalDiffTime))
+    -- ^ Time the created mint/burn transaction is valid until.
+    } deriving (Eq, Generic, Show)
+
+instance DecodeAddress n => FromJSON (PostMintBurnAssetData n) where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+
+instance EncodeAddress n => ToJSON (PostMintBurnAssetData n) where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+-- | Core minting and burning request information.
+--
+-- Assets are minted and burned under a "policy". The policy defines under what
+-- circumstances a token may be minted and burned. The typical policy is "A
+-- token may be minted and burned if signature 's' witnesses the transaction,
+-- for some signature 's'". This is the only type of policy supported by the
+-- cardano-wallet API at the moment. Because cardano-wallet manages the keys of
+-- the user, we ask the user not for a specific signature, but rather for a key
+-- derivation index, which we use to derive the signature to construct the
+-- policy with.
+data ApiMintBurnData (n :: NetworkDiscriminant) = ApiMintBurnData
+    { monetaryPolicyIndex :: !(Maybe (ApiT DerivationIndex))
+    -- ^ The key derivation index to use to construct the policy.
+    , assetName           :: !(ApiT W.TokenName)
+    -- ^ The name of the asset to mint/burn.
+    , operations          :: !(NonEmpty (ApiMintBurnOperation n))
+    -- ^ The set of minting and burning operations to perform.
```

Comment:
This doesn't seem quite right to me. There could be contradictory elements in the operations list.

What about this?
```suggestion
    , assetNames :: !(Map (ApiT W.TokenName) (ApiMintBurnOperation n))
    -- ^ A mint/burn amount for assets of the policy script
```

Reply:
It's not quite what it seems. The operations contain no reference to what they are operating on (it's either `mint 3` or `burn 4`, there is no reference to the asset).

The caller can choose one monetary policy index, one asset name, then any number of mint/burn operations for that asset.

I should probably allow the user to operate on multiple assets:

```
 data PostMintBurnAssetData (n :: NetworkDiscriminant) = PostMintBurnAssetData
-    { mintBurn   :: !(ApiMintBurnData n)
-    { mintBurn   :: !(NonEmpty (ApiMintBurnData n)) 
     -- ^ Minting and burning request.
```

And then offer a singular operation for each asset, sorta like it's done in Tx construction:

```
 data ApiMintBurnData (n :: NetworkDiscriminant) = ApiMintBurnData
     { monetaryPolicyIndex :: !(Maybe (ApiT DerivationIndex))
     -- ^ The key derivation index to use to construct the policy.
     , assetName           :: !(ApiT W.TokenName)
     -- ^ The name of the asset to mint/burn.
-    , operations          :: !(NonEmpty (ApiMintBurnOperation n))
+    , operation           :: !(ApiMintBurnOperation n)
     } deriving (Eq, Generic, Show)
```

Reply:
I've made this change in e7adcde7f.

Reply:
Oh... But if the user is allowed to mint multiple assets in one API call I'll have to rework the response... a sec.

Reply:
Addressed in 619ba48.